src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,344 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Texl.Builtins;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class TexlFunctionSet
+    {
+        // Dictionary key: function.Name
+        private Dictionary<string, List<TexlFunction>> _functions;
+
+        // Dictionary key: function.LocaleInvariantName
+        private Dictionary<string, List<TexlFunction>> _functionsInvariant;
+
+        // Dictionary key: function.Namespace
+        private Dictionary<DPath, List<TexlFunction>> _namespaces;
+
+        // List of all function.GetRequiredEnumNames()
+        private List<string> _enums;
+
+        // Count of functions
+        internal int _count;
+
+        internal Dictionary<string, List<TexlFunction>>.KeyCollection FunctionNames => _functions.Keys;
+
+        internal Dictionary<string, List<TexlFunction>>.KeyCollection InvariantFunctionNames => _functionsInvariant.Keys;
+
+        internal Dictionary<DPath, List<TexlFunction>>.KeyCollection Namespaces => _namespaces.Keys;
+
+        internal IEnumerable<TexlFunction> WithName(string name) => WithNameInternal(name);
+
+        private List<TexlFunction> WithNameInternal(string name) => _functions.ContainsKey(name) ? _functions[name] : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithName(string name, DPath ns) => _functions.ContainsKey(name) ? new List<TexlFunction>(_functions[name].Where(f => f.Namespace == ns)) : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithInvariantName(string name) => WithInvariantNameInternal(name);
+
+        private List<TexlFunction> WithInvariantNameInternal(string name) => _functionsInvariant.ContainsKey(name) ? _functionsInvariant[name] : new List<TexlFunction>();
```

Comment:
>_functionsInvariant.ContainsKey(name) ? _functionsInvariant[name] [](http://example.com/codeflow?start=77&length=65)

nit; use TryGetValue  - single lookup. rather than the double lookup